exchanges/subscription/template_test.go

```diff
@@ -84,13 +84,18 @@ func TestExpandTemplates(t *testing.T) {
 	}
 	equalLists(t, exp, got)
 
+	// Users can specify pairs which aren't available, even across diverse assets
```

Comment:
This happened about two months ago; I'll be trying to pull highlights from old messages, but it might still be a bit of a mess.

- If exchanges/subscription/list.go function assetPairs is passed a list of subscriptions without any assets, it silently passes an empty assetPairs struct onwards.
- This eventually goes to exchange/subscription/template.go function expandTemplate which checks whether the pairs in the subscription request are in the assetPairs struct. Since that struct's empty, this errors out and rewinds.
- This causes an issue, since one of the Coinbase websocket channels doesn't take in an asset as a parameter; it returns results for all assets when a pair is passed in. A request for "BTC-USD" will return all spot, perpetual futures, and expiring futures associated with those two currencies, including pairs like "BTC-PERP-INTX" and "BIT-28JUN24-CDE".
- One could attempt to fix this by passing asset.All as the asset type, but that causes an issue, since some asset types on the exchange don't have pairs that are literally identical to "BTC-USD" in their enabled list, yet the exchange will do this "fanning out" from BTC-USD to BTC-PERP-INTX and the like for us.

Let me know if anything needs clarification, or went unexplained.

Reply:
In terms of the code in `template.go` `expandTemplate`, the code that was changed in #1635 and previously, is not even hit. You have set up CBP subscriptions to not pass in pairs in the first place, they are instead retrieved inside `expandTemplate` in the `for i, assetChannels := range assetRecords {` loop. 
So the previous code of
```go
if len(s.Pairs) != 0 {
		for a, pairs := range subCtx.AssetPairs {
			if err := pairs.ContainsAll(s.Pairs, true); err != nil { //nolint:govet // Shadow, or gocritic will complain sloppyReassign
				return nil, err
			}
			subCtx.AssetPairs[a] = s.Pairs
		}
	}
```
Would never be hit since `s.Pairs` is `nil` for `user` subscription (and all subscriptions you are generating).

I'm definitely fine with `"user"` sub being without an asset, but I'm still left confused by the code change and the comment since it appears as though it doesn't apply.

Plus even if you were passing in a pair with the `user` subscription, getting the _results_ with all matching BTC-USD would not have to do with the subscription itself but based on what data is returned by coinbasepro websocket after the subscription and handled in coinbasepro_websocket.go -> `wsHandleData`, right?

Perhaps there has been other bugfixes that have resolved the issue you speak of. Or this is all masked by the config subscription bug I just mentioned. We can talk later 🤙

Reply:
Alright, I've figured some stuff out:

- Config overwrite was due to the checkSubscriptions function. It was treating "heartbeat" as outdated, despite that being our engine's internal name for that type of channel (and hence, the name that gets saved in the config). Removing that part of the check has fixed that issue.
- Passing in asset.All for the user channel doesn't cause any errors with the exchange or engine, but is a bit weird and might cause some redundancy. It will send in futures contracts that will be covered by the names of spot pairs, but the exchange doesn't seem to care.
- CBP subscriptions can still be passed pairs. This can be done by altering the config file to include specific pairs for specific subscriptions, or by calling the subscription function during testing (as can be seen in TestWsAuth). 
- When the user channel is passed pairs, PR 1635 is needed to prevent it from erroring out.
- Passing in no asset for the user channel results in issues. If pairs are passed as well, it will error out without PR 1635. If pairs aren't passed, then expandTemplate won't automatically include any, causing no subscription to be sent.
- I'm not fully sure with what you mean by the "Plus even if you were passing..." paragraph. If you just meant to say "The actual content of the initial description doesn't really matter when handling it", that's true; PR 1635 is only relevant for the initial subscription.

Reply:
Thank you for your detailed explanation! I learned about the `pairs` property of subscriptions and that you support it properly. I can now say I've experienced what you describe:
`[ERROR] | WEBSOCKET | 05/11/2024 13:46:19 | CoinbasePro websocket: subscription failure, account all BTC/USD: BTC-USD pair not contained in available pairs`
You can revert the commit issues with lint, the old code re-introduction and the `{Enabled: false` for subs 
🎉